build: Add ability to build qt in depends with -stdlib=libc++

hebasto opened the pull request:
This PR makes possible to build the `qt` package in depends against `libc++` for x86_64 platform.

Fixes #22344.

Required for #22815.

Also this PR [fixes](https://github.com/bitcoin/bitcoin/pull/23060#discussion_r716077050) the `[no wallet] [bionic]` task on CI:
- on master (a8bbd4cc819633ec50ed0f763b6a75330ae055fb), https://api.cirrus-ci.com/v1/task/5558609250615296/logs/ci.log:
```
Options used to compile and link:
  external signer = yes
  multiprocess    = no
  with libs       = yes
  with wallet     = no
  with gui / qt   = no
```
- this PR, https://api.cirrus-ci.com/v1/task/5502605561430016/logs/ci.log:
```
Options used to compile and link:
  external signer = yes
  multiprocess    = no
  with libs       = yes
  with wallet     = no
  with gui / qt   = yes
```

hebasto committed:
build: Add ability to build qt in depends with -stdlib=libc++

jarolrod commented:
GUIX hashes mine match @hebasto:
```
find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum

db06398ab9b1be8f1d28c19c2a07fb60ca5ca499136258e5de9388af2ac795cb  guix-build-7ae424b899cf/output/aarch64-linux-gnu/SHA256SUMS.part
f700183e5fcb96e9cea08089e657dc5ed9c23dab92a7cf7952a2c7db654f1883  guix-build-7ae424b899cf/output/aarch64-linux-gnu/bitcoin-7ae424b899cf-aarch64-linux-gnu-debug.tar.gz
ffd040f434d84f5c1a46eda650f65ade7ccac0f5719a789ade4aace267c10dc4  guix-build-7ae424b899cf/output/aarch64-linux-gnu/bitcoin-7ae424b899cf-aarch64-linux-gnu.tar.gz
a2fe9a28b19aef232c55fcd35312b0f9d6790a25e9d5a872abc28b5cd324e323  guix-build-7ae424b899cf/output/arm-linux-gnueabihf/SHA256SUMS.part
79746b31031f5950190a472f9a91583d9ff4403cd856677269bcc138b3576029  guix-build-7ae424b899cf/output/arm-linux-gnueabihf/bitcoin-7ae424b899cf-arm-linux-gnueabihf-debug.tar.gz
2615275085c93cd089f1f500ce4253f75d4c58bf24981b53b1c85206a634ec6e  guix-build-7ae424b899cf/output/arm-linux-gnueabihf/bitcoin-7ae424b899cf-arm-linux-gnueabihf.tar.gz
0ad134b4125aa88d782c7f3a80759215653198ff05a237244bc596eeb3634c1c  guix-build-7ae424b899cf/output/dist-archive/bitcoin-7ae424b899cf.tar.gz
4e9ddbb7cc2ac67d1cbf79664c33ba8d3b9f87cc6c1eac5249b6210c15130c2f  guix-build-7ae424b899cf/output/powerpc64-linux-gnu/SHA256SUMS.part
5b4c2beb03ad62db8de988816b5a510153bc2fdece83c030d1b31a84aa5d909e  guix-build-7ae424b899cf/output/powerpc64-linux-gnu/bitcoin-7ae424b899cf-powerpc64-linux-gnu-debug.tar.gz
4a7f5af8c45e169e19cd0c11bc16a348d88101199406b62541e6592a72b5081c  guix-build-7ae424b899cf/output/powerpc64-linux-gnu/bitcoin-7ae424b899cf-powerpc64-linux-gnu.tar.gz
e502aac9fda9c243ba6d9811f39b4c05e63be3f26ceffe8258baa82a1392aa48  guix-build-7ae424b899cf/output/powerpc64le-linux-gnu/SHA256SUMS.part
031dc8a8013be209535bfd2ebaf08bfdedb8144f8112471eb4f2bcb52dea4f85  guix-build-7ae424b899cf/output/powerpc64le-linux-gnu/bitcoin-7ae424b899cf-powerpc64le-linux-gnu-debug.tar.gz
198a468ef0ef8a123fedb5fe71687111eebfaf9b7802683b0f95b986d417d38f  guix-build-7ae424b899cf/output/powerpc64le-linux-gnu/bitcoin-7ae424b899cf-powerpc64le-linux-gnu.tar.gz
ed7bfa400f02db88b8be10c48b1dd711bd39917daba0a59a431b18074c1b2d37  guix-build-7ae424b899cf/output/riscv64-linux-gnu/SHA256SUMS.part
edf5c552492e0795a78dd4f7f053f0d57ad2fcae80db9e654da7c5a79c941c42  guix-build-7ae424b899cf/output/riscv64-linux-gnu/bitcoin-7ae424b899cf-riscv64-linux-gnu-debug.tar.gz
883bb26954b56bfe31c952bdd8b40541110f53a2181d719b321c49da01738e53  guix-build-7ae424b899cf/output/riscv64-linux-gnu/bitcoin-7ae424b899cf-riscv64-linux-gnu.tar.gz
9b01fc0dffcdde81ca4f87197652b6d499b9482ea7d87a496e71bf2bc9fdb470  guix-build-7ae424b899cf/output/x86_64-apple-darwin18/SHA256SUMS.part
c10750e2271074ea735d5b908f5ec206dfbf6d4fbe64bca0ae31caa8ae5e942d  guix-build-7ae424b899cf/output/x86_64-apple-darwin18/bitcoin-7ae424b899cf-osx-unsigned.dmg
9bb167d2f957baa1b1150823ba7064be6ff47459dfe1aa3821aa4e1a23775840  guix-build-7ae424b899cf/output/x86_64-apple-darwin18/bitcoin-7ae424b899cf-osx-unsigned.tar.gz
2ed89ce2ddfdbee3240c3653c074955ea1a800fb6a386c34edfbca6394dd1754  guix-build-7ae424b899cf/output/x86_64-apple-darwin18/bitcoin-7ae424b899cf-osx64.tar.gz
668afc33cbf9791933a92727f360d7d553a5de36d12162af2a9a25c256847fc7  guix-build-7ae424b899cf/output/x86_64-linux-gnu/SHA256SUMS.part
0f913f662f5d16d499bd60177221b8ab02c2ca3fff28b56a28da9c7a20480f91  guix-build-7ae424b899cf/output/x86_64-linux-gnu/bitcoin-7ae424b899cf-x86_64-linux-gnu-debug.tar.gz
cc1a1a1bcb621b20bb982fa2b8b79a8f94358c34d8dd7d474ecbfff9db72ff14  guix-build-7ae424b899cf/output/x86_64-linux-gnu/bitcoin-7ae424b899cf-x86_64-linux-gnu.tar.gz
57a4a6e420c4f8721e4981a306c0f94f2ca6578f362c9fb7e5a4d4378027d43c  guix-build-7ae424b899cf/output/x86_64-w64-mingw32/SHA256SUMS.part
120c7df398b70556bff009afb9e3a96780ba6b46d59f7af98b0e3140f8ff1089  guix-build-7ae424b899cf/output/x86_64-w64-mingw32/bitcoin-7ae424b899cf-win-unsigned.tar.gz
3f24b0301b20de690516cfe734a84c83e05bea64c31d767770e239466da06586  guix-build-7ae424b899cf/output/x86_64-w64-mingw32/bitcoin-7ae424b899cf-win64-debug.zip
5eb6564689adac44c63c800586db0740515786a244ba506658f988ec5711f759  guix-build-7ae424b899cf/output/x86_64-w64-mingw32/bitcoin-7ae424b899cf-win64-setup-unsigned.exe
4ac3c715e8977fcfd6b9ae50a54f3b8080e6bb69367960e05f53551c716ec965  guix-build-7ae424b899cf/output/x86_64-w64-mingw32/bitcoin-7ae424b899cf-win64.zip
```

fanquake commented on depends/packages/qt.mk:
```diff
@@ -140,7 +140,11 @@ $(package)_config_opts_linux += -no-feature-vulkan
 $(package)_config_opts_linux += -dbus-runtime
 $(package)_config_opts_arm_linux += -platform linux-g++ -xplatform bitcoin-linux-g++
 $(package)_config_opts_i686_linux  = -xplatform linux-g++-32
+ifneq (,$(findstring clang++ -stdlib=libc++,$($(1)_cxx)))
```
As mentioned in the comment from https://github.com/bitcoin/bitcoin/pull/22815#discussion_r697683792:

> Looking only at the argument (which should directly follow the compiler name) looks kind of fragile and will not catch every case where someone links against libc++.

To expand on that, as-is this also wont work when you're using a versioned clang, i.e `clang++-12`, which we often do in our CIs.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/22814#issuecomment-928891018).

hebasto commented:
Updated 7ae424b899cf5fb32f0a0bba487ea54e53ab276f -> 33796a964a8c605ce482959456d1ca78638340d9 ([pr22814.01](https://github.com/hebasto/bitcoin/commits/pr22814.01) -> [pr22814.02](https://github.com/hebasto/bitcoin/commits/pr22814.02)):

- addressed https://github.com/bitcoin/bitcoin/pull/22814#discussion_r717242412
- rebased on top of the #23060 (see https://github.com/bitcoin/bitcoin/pull/23060#discussion_r716077050)

hebasto commented:
#### Guix builds:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum
31a154cc2f48e8ca97f653f654d468df9f9e576d04e4e6108d78cbd4e861e47c  guix-build-33796a964a8c/output/aarch64-linux-gnu/SHA256SUMS.part
d86132ea582f2bd4f9c628ac7df4a2aa6e0c3e64a48bf0a16cd1c3293f8b2b27  guix-build-33796a964a8c/output/aarch64-linux-gnu/bitcoin-33796a964a8c-aarch64-linux-gnu-debug.tar.gz
d43e49a3443413f72383c4327f402de787af0b7bbc7d8e661e651137bd636f73  guix-build-33796a964a8c/output/aarch64-linux-gnu/bitcoin-33796a964a8c-aarch64-linux-gnu.tar.gz
673b67b092e19486d5fede917a5236f0f5320d70664903152717ecefbe120b05  guix-build-33796a964a8c/output/arm-linux-gnueabihf/SHA256SUMS.part
fbc7f256d71d0a30f18319cc68a52c32d0f192b1a3974dc34e4e2cb1a508044a  guix-build-33796a964a8c/output/arm-linux-gnueabihf/bitcoin-33796a964a8c-arm-linux-gnueabihf-debug.tar.gz
14bbca1070e063a7581f91cfaec15b5002aa07d26e6f7c111527c3865cbd7325  guix-build-33796a964a8c/output/arm-linux-gnueabihf/bitcoin-33796a964a8c-arm-linux-gnueabihf.tar.gz
caed4f216af6f9917bde8d944165d27848227bcbccca11838c7634d6dcb7dbe5  guix-build-33796a964a8c/output/dist-archive/bitcoin-33796a964a8c.tar.gz
0db4e2338b1fa32661b87623e0fd68a8a7d8f69ce9cbfb5ed9f1937a5f50b19f  guix-build-33796a964a8c/output/powerpc64-linux-gnu/SHA256SUMS.part
a427c3af3cdfd25e4e4c4ffbea0c7e1267d21a6e7bbab827568e9c50166f571f  guix-build-33796a964a8c/output/powerpc64-linux-gnu/bitcoin-33796a964a8c-powerpc64-linux-gnu-debug.tar.gz
9569e3390fb92a15316373f6808ed0bce80ca2fd0aed4a3fa25498bf4ecca076  guix-build-33796a964a8c/output/powerpc64-linux-gnu/bitcoin-33796a964a8c-powerpc64-linux-gnu.tar.gz
8e025033de78cb51c74ed1cfc0e3af560ec1a8a41d553ffac84475532073461a  guix-build-33796a964a8c/output/powerpc64le-linux-gnu/SHA256SUMS.part
9da7c9d12736f18c1f8bb0c58602a7aa4e6053e5bed61a2bb6fe09b5418e0030  guix-build-33796a964a8c/output/powerpc64le-linux-gnu/bitcoin-33796a964a8c-powerpc64le-linux-gnu-debug.tar.gz
30020d531a2ab4bd5278d7c399ee23eadf3c6e2479df0a79dcefb750d81d99cf  guix-build-33796a964a8c/output/powerpc64le-linux-gnu/bitcoin-33796a964a8c-powerpc64le-linux-gnu.tar.gz
1c0bf18ef8dc592ed5864f121cca0ce0ca4b35710bcdc9fca8ce5fceec4bf7e5  guix-build-33796a964a8c/output/riscv64-linux-gnu/SHA256SUMS.part
0f7bbf8d0c13b9a519883f8a4697324eb410940fe83d03fb69da485edee66ac1  guix-build-33796a964a8c/output/riscv64-linux-gnu/bitcoin-33796a964a8c-riscv64-linux-gnu-debug.tar.gz
cc68b0024c038b4eaa831ba261d58a0204886dd5ed6d0d227f450b01da2d77d5  guix-build-33796a964a8c/output/riscv64-linux-gnu/bitcoin-33796a964a8c-riscv64-linux-gnu.tar.gz
1876ed8ba6f21810fe990ab4bfa167506e568716c242385bf7d7dde130a2dcf4  guix-build-33796a964a8c/output/x86_64-apple-darwin19/SHA256SUMS.part
c646278110ed3a1e13935e549a8bc94f7aaa61e2eccdf9d9bf26a34e1b7fe3c6  guix-build-33796a964a8c/output/x86_64-apple-darwin19/bitcoin-33796a964a8c-osx-unsigned.dmg
168df829ba479e3d55500e825c22e4db2d464c238176475c304c923b72201cd2  guix-build-33796a964a8c/output/x86_64-apple-darwin19/bitcoin-33796a964a8c-osx-unsigned.tar.gz
8d35196f6010aafa93a29e11a9b25c894eef32162e897bef98beb60d452fc6d7  guix-build-33796a964a8c/output/x86_64-apple-darwin19/bitcoin-33796a964a8c-osx64.tar.gz
41560bde25a0c8606e91e7166a8ad5239b86066692ec588dc4537f970e0acee4  guix-build-33796a964a8c/output/x86_64-linux-gnu/SHA256SUMS.part
9197132a68e6b2bba14a706f9e758e7399d62c5ed9473ef85d13b9b2fd98c2fc  guix-build-33796a964a8c/output/x86_64-linux-gnu/bitcoin-33796a964a8c-x86_64-linux-gnu-debug.tar.gz
8753052b0996e3411d0ac221dc4601cdb2a7b9f93278573d13dbb177a17b7d33  guix-build-33796a964a8c/output/x86_64-linux-gnu/bitcoin-33796a964a8c-x86_64-linux-gnu.tar.gz
88dfef76cfc0ee47635328537e64af72a066ad76f9bd7dac33364d02c2132678  guix-build-33796a964a8c/output/x86_64-w64-mingw32/SHA256SUMS.part
d746acda77584ec0f2f41a016a94b3d192d4f1087baf5706970743d509a5531f  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win-unsigned.tar.gz
f757730a7f6a327915e1b07f73927454c9870f675ee00b20f9d3204a8c8a1d15  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win64-debug.zip
0c211ecd900ea5de9d552b2ae2c19665c9b48e7dbab688f54c46c8e891f62307  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win64-setup-unsigned.exe
4b3fdc40d3028c6e1663f498ed04e5ead39922986b7928c8e8966c2664a92b31  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win64.zip
```

hebasto commented:
Also this PR [fixes](https://github.com/bitcoin/bitcoin/pull/23060#discussion_r716077050) the `[no wallet] [bionic]` task on CI:
- on master (a8bbd4cc819633ec50ed0f763b6a75330ae055fb), https://api.cirrus-ci.com/v1/task/5558609250615296/logs/ci.log:
```
Options used to compile and link:
  external signer = yes
  multiprocess    = no
  with libs       = yes
  with wallet     = no
  with gui / qt   = no
```
- this PR, https://api.cirrus-ci.com/v1/task/5502605561430016/logs/ci.log:
```
Options used to compile and link:
  external signer = yes
  multiprocess    = no
  with libs       = yes
  with wallet     = no
  with gui / qt   = yes
```

MarcoFalke commented:
Concept ACK

jarolrod commented:
contributing guix hashes, mine match @hebasto second round of hashes:

```
find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum

31a154cc2f48e8ca97f653f654d468df9f9e576d04e4e6108d78cbd4e861e47c  guix-build-33796a964a8c/output/aarch64-linux-gnu/SHA256SUMS.part
d86132ea582f2bd4f9c628ac7df4a2aa6e0c3e64a48bf0a16cd1c3293f8b2b27  guix-build-33796a964a8c/output/aarch64-linux-gnu/bitcoin-33796a964a8c-aarch64-linux-gnu-debug.tar.gz
d43e49a3443413f72383c4327f402de787af0b7bbc7d8e661e651137bd636f73  guix-build-33796a964a8c/output/aarch64-linux-gnu/bitcoin-33796a964a8c-aarch64-linux-gnu.tar.gz
673b67b092e19486d5fede917a5236f0f5320d70664903152717ecefbe120b05  guix-build-33796a964a8c/output/arm-linux-gnueabihf/SHA256SUMS.part
fbc7f256d71d0a30f18319cc68a52c32d0f192b1a3974dc34e4e2cb1a508044a  guix-build-33796a964a8c/output/arm-linux-gnueabihf/bitcoin-33796a964a8c-arm-linux-gnueabihf-debug.tar.gz
14bbca1070e063a7581f91cfaec15b5002aa07d26e6f7c111527c3865cbd7325  guix-build-33796a964a8c/output/arm-linux-gnueabihf/bitcoin-33796a964a8c-arm-linux-gnueabihf.tar.gz
caed4f216af6f9917bde8d944165d27848227bcbccca11838c7634d6dcb7dbe5  guix-build-33796a964a8c/output/dist-archive/bitcoin-33796a964a8c.tar.gz
0db4e2338b1fa32661b87623e0fd68a8a7d8f69ce9cbfb5ed9f1937a5f50b19f  guix-build-33796a964a8c/output/powerpc64-linux-gnu/SHA256SUMS.part
a427c3af3cdfd25e4e4c4ffbea0c7e1267d21a6e7bbab827568e9c50166f571f  guix-build-33796a964a8c/output/powerpc64-linux-gnu/bitcoin-33796a964a8c-powerpc64-linux-gnu-debug.tar.gz
9569e3390fb92a15316373f6808ed0bce80ca2fd0aed4a3fa25498bf4ecca076  guix-build-33796a964a8c/output/powerpc64-linux-gnu/bitcoin-33796a964a8c-powerpc64-linux-gnu.tar.gz
8e025033de78cb51c74ed1cfc0e3af560ec1a8a41d553ffac84475532073461a  guix-build-33796a964a8c/output/powerpc64le-linux-gnu/SHA256SUMS.part
9da7c9d12736f18c1f8bb0c58602a7aa4e6053e5bed61a2bb6fe09b5418e0030  guix-build-33796a964a8c/output/powerpc64le-linux-gnu/bitcoin-33796a964a8c-powerpc64le-linux-gnu-debug.tar.gz
30020d531a2ab4bd5278d7c399ee23eadf3c6e2479df0a79dcefb750d81d99cf  guix-build-33796a964a8c/output/powerpc64le-linux-gnu/bitcoin-33796a964a8c-powerpc64le-linux-gnu.tar.gz
1c0bf18ef8dc592ed5864f121cca0ce0ca4b35710bcdc9fca8ce5fceec4bf7e5  guix-build-33796a964a8c/output/riscv64-linux-gnu/SHA256SUMS.part
0f7bbf8d0c13b9a519883f8a4697324eb410940fe83d03fb69da485edee66ac1  guix-build-33796a964a8c/output/riscv64-linux-gnu/bitcoin-33796a964a8c-riscv64-linux-gnu-debug.tar.gz
cc68b0024c038b4eaa831ba261d58a0204886dd5ed6d0d227f450b01da2d77d5  guix-build-33796a964a8c/output/riscv64-linux-gnu/bitcoin-33796a964a8c-riscv64-linux-gnu.tar.gz
1876ed8ba6f21810fe990ab4bfa167506e568716c242385bf7d7dde130a2dcf4  guix-build-33796a964a8c/output/x86_64-apple-darwin19/SHA256SUMS.part
c646278110ed3a1e13935e549a8bc94f7aaa61e2eccdf9d9bf26a34e1b7fe3c6  guix-build-33796a964a8c/output/x86_64-apple-darwin19/bitcoin-33796a964a8c-osx-unsigned.dmg
168df829ba479e3d55500e825c22e4db2d464c238176475c304c923b72201cd2  guix-build-33796a964a8c/output/x86_64-apple-darwin19/bitcoin-33796a964a8c-osx-unsigned.tar.gz
8d35196f6010aafa93a29e11a9b25c894eef32162e897bef98beb60d452fc6d7  guix-build-33796a964a8c/output/x86_64-apple-darwin19/bitcoin-33796a964a8c-osx64.tar.gz
41560bde25a0c8606e91e7166a8ad5239b86066692ec588dc4537f970e0acee4  guix-build-33796a964a8c/output/x86_64-linux-gnu/SHA256SUMS.part
9197132a68e6b2bba14a706f9e758e7399d62c5ed9473ef85d13b9b2fd98c2fc  guix-build-33796a964a8c/output/x86_64-linux-gnu/bitcoin-33796a964a8c-x86_64-linux-gnu-debug.tar.gz
8753052b0996e3411d0ac221dc4601cdb2a7b9f93278573d13dbb177a17b7d33  guix-build-33796a964a8c/output/x86_64-linux-gnu/bitcoin-33796a964a8c-x86_64-linux-gnu.tar.gz
88dfef76cfc0ee47635328537e64af72a066ad76f9bd7dac33364d02c2132678  guix-build-33796a964a8c/output/x86_64-w64-mingw32/SHA256SUMS.part
d746acda77584ec0f2f41a016a94b3d192d4f1087baf5706970743d509a5531f  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win-unsigned.tar.gz
f757730a7f6a327915e1b07f73927454c9870f675ee00b20f9d3204a8c8a1d15  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win64-debug.zip
0c211ecd900ea5de9d552b2ae2c19665c9b48e7dbab688f54c46c8e891f62307  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win64-setup-unsigned.exe
4b3fdc40d3028c6e1663f498ed04e5ead39922986b7928c8e8966c2664a92b31  guix-build-33796a964a8c/output/x86_64-w64-mingw32/bitcoin-33796a964a8c-win64.zip
```

fanquake reviewed (approved):
ACK 33796a964a8c605ce482959456d1ca78638340d9 - While this sort of string matching is fragile, I think the risk of this causing any actual issues is low. 

Anyone who was previously building with `-stdlib=libc++` in their CXXFLAGS (us in CI) now gets a working Qt build. 

Anywhere that was using a Clang which uses `-stdlib=libc++` by default (i.e no passing in flags), will get the same (`xplatform linux-g++-64`) build. If they start passing `-stdlib=libc++` they'll, if anything, get a "more correct" Qt build.

Anyone who is building with GCC / anything else / doesn't care about this, still gets the same `-xplatform linux-g++-64` build.

Note that this is also scoped to Linux, so can't cause issues with the macOS builds which use `libc++`.
